### PR TITLE
Simple log package

### PR DESCRIPTION
**Pull request by hugelgupf**
WDYT?

### CODE REVIEW

File: pkg/log/log.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
+// Copyright 2017 the u-root Authors. All rights reserved.
```
Comment (rminnich) [inline pkg/log/log.go]: not for now, but is this a good place to define a way to log 'progress', including the basic 'print progress on a SIGHUP'?

File: pkg/log/log.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+// Printf emits the given log string with time prefixed on the log writer.
+// Arguments are handled as in fmt.Printf.
+func Printf(format string, args ...interface{}) {
+	if verbosity {
+		Print(fmt.Sprintf(format, args...))
+	}
+}
+
+// PrintObj emits a json-encoded obj on the log writer.
+func PrintObj(obj interface{}) {
```
Comment (rminnich) [inline pkg/log/log.go]: oooh, nice

File: pkg/log/log.go (hunk 1 above)
Comment (hugelgupf) [inline pkg/log/log.go]: it will be, yeah.

**Comment by hugelgupf** [comment]
This is getting some more commits, PTAL

**Comment by tpepper** [comment]
I'm severely late to the conversation, but for the first commit I'm curious to see a problem statement.  It seems like every big golang based project tries writing its own logging system.

**Comment by hugelgupf** [comment]
You're right, many go-based projects do, and it's kind of a pain.

If you look through our existing code (e.g. the cpio command, or the dhclient command), you will see that we define these "debug" functions all over the place, with different command-line flags to enable logging or not. All I wanted to be able to do is remove the usage of `debug` and standardize to a set of flags used to enable logging -- without having to copy and paste the solution into every command we have. This is the simplest thing I could come up with.

File: cmds/builtin/builtin.go [hunk 3]
```diff
@@ -116,13 +111,12 @@ func main() {
 	// There is no FIXME
 	if false {
```
Comment (xtuc) [inline cmds/builtin/builtin.go]: Could we get rid of this?

You could make an issue instead containing the code.

**Comment by rminnich** [comment]
Tom, I think the problem is that the go based logging is not quite right, but nobody has a better idea. Sort of like the go options processing :-)

I got the idea of
var debug = func(string, ...interface{}) {}
from some of Russ's code. 

I don't see the need to capture all that behavior in a logging package. To me it's almost like capturing stuff like
i = 5
in a function like
init(&i, 5); // yes, I've really seen this in C code.

Also, the debug = func(){} approach is nice as it leaves it possible to use anything that matches that signature as a logger, which behavior is tough to capture in a logging package.

On a different note, sometimes the debug = func(){} approach  is a very bad way to go, since even for the empty function all arguments are evaluated and in the case of long-lived network protocols the performance impact can be substantial. It's why you also see stuff like if (mask & BIT) log.Printf(etc)

capturing all the behaviour of all possible logging uses is tough, and it's one reason nobody's every really gotten it totally right.

**Comment by rminnich** [comment]
One other note. Log packages have always proven to be hard *if you try to support all possible uses*. They tend to turn into giant option-filled bowls of spaghetti. So there's a real trick to it, and, I'll warn again that there are few if any examples that work universally well.

**Comment by tpepper** [comment]
This all makes sense and is expected.  I just want to see it articulated in the commit history for posterity, eg:  "This is not a general logging mechanism.  It is meant for capturing ${this and that} for debug scenarios to replace the mishmash of debug print implementations currently in the code.  Use normal prints where you normally would for expected stdout/stderr, but using these debug ones is the standard u-root way for anywhere you're expected to need to debug ${this and that}."

**Comment by hugelgupf** [comment]
I'm gonna deal with this at a later time - closing it for now.

**Pull request closed without merging**